test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/DefaultHtmlGeneratorTest.cs

```diff
@@ -190,6 +191,103 @@ public void GenerateTextArea_WithNullExpression_WithNameAttribute_DoesNotThrow()
             Assert.Equal(expected, attribute.Value);
         }
 
+        [Theory]
+        [MemberData(nameof(GetMaxLengthAttributeRenderingTestsData))]
+        public void Generate_RendersMaxLengthAttribute_AsRequested(
+            string expression,
+            int expectedValue,
+            Action<Dictionary<string, object>> customizeHtmlAttributes,
+            bool maxLengthShouldBeRendered,
+            Func<IHtmlGenerator, ViewContext, ModelExplorer, Dictionary<string, object>, string, TagBuilder> getTagBuilder)
+        {
+            // Arrange
+            var metadataProvider = new TestModelMetadataProvider();
+            var htmlGenerator = GetGenerator(metadataProvider);
+            var viewContext = GetViewContext<ModelWithMaxLengthMetadata>(model: null, metadataProvider: metadataProvider, new HtmlHelperOptions { MaxLengthAttributeRenderingEnabled = true });
+            var modelMetadata = metadataProvider.GetMetadataForProperty(typeof(ModelWithMaxLengthMetadata), expression);
+            var modelExplorer = new ModelExplorer(metadataProvider, modelMetadata, null);
+            var htmlAttributes = new Dictionary<string, object>
+            {
+                { "name", "testElement" },
+            };
+
+            customizeHtmlAttributes?.Invoke(htmlAttributes);
+
+            // Act
+            var tagBuilder = getTagBuilder(htmlGenerator, viewContext, modelExplorer, htmlAttributes, expression);
```

Comment:
Wow, never seen a test where the whole `// Act` is passed in. We should probably have separate tests for `GenerateTextArea(...)` et cetera. (Those tests could share a helper or two for `// Arrange` and the`TheoryData`.)

Reply:
That will mean artificially bloating the code. This way it's super compact.
@pranavkm, thoughts regarding this? If both of you think the same way, I'll change.

Reply:
Per offline conversation, will split apart the tests

Reply:
You haven't split the tests sufficiently to make it possible to see what the test is testing. Get rid of the `Func<T...>`, these helpers, and repeat the `// Arrange` bits in the distinct tests.